index.d.ts

```diff
@@ -1,8 +1,44 @@
-export function format(input: object, options?: {
-    abbreviate: boolean;
-    appendCountry: boolean;
-    cleanupPostcode: boolean;
-    countryCode: string;
-    fallbackCountryCode: string;
-    output: 'string' | 'array';
-}): string | string[];
+// TODO: add aliases
+/** @see [here](https://github.com/OpenCageData/address-formatting/blob/master/conf/components.yaml) */
+type Input = Partial<
+  Record<
+    | 'houseNumber'
+    | 'house'
+    | 'road'
+    | 'hamlet'
+    | 'village'
+    | 'neighbourhood'
+    | 'postalCity'
+    | 'city'
+    | 'municipality'
+    | 'county'
+    | 'stateDistrict'
+    | 'postcode'
+    | 'state'
+    | 'region'
+    | 'island'
+    | 'archipelago'
+    | 'country'
+    | 'countryCode'
+    | 'continent',
+    string
+  >
+>;
+
+interface CommonOptions {
+  abbreviate?: boolean;
+  appendCountry?: boolean;
+  cleanupPostcode?: boolean;
+  countryCode?: string;
+  fallbackCountryCode?: string;
+}
+
+export function format(
+  input: Input,
+  options?: CommonOptions & { output?: 'string' },
```

Comment:
Any reason for having output specially called out here? Why is it not in the `CommonOptions`?

Reply:
Yes, this function should be overloaded and the returning type should be based on `output` param. E.g.

```ts
const formattedAsString = format(someInput, { output: 'string'});

const formattedAsArray = format(someInput, {output: 'string'});
```

Reply:
in the 2nd example, output should be `'array'`. @JirkaChadima with function overloading you can make sure that the output type is dependent on the input. Also, since string is default, omitting the output will also yield a `string` type.

Reply:
🤦🏻 pardon